### IDs:
Insert yours IDs to the cell below

ID #1:310116165

ID #2:308564293


## Read the following instructions carefully:

1. This jupyter notebook contains all the step by step instructions needed for this exercise.
1. You are free to add cells.
1. Write your functions and your answers in this jupyter notebook only.
1. Answers to theoretical questions should be written in **markdown cells (with $\LaTeX$ support)**.
1. Submit this jupyter notebook only using your ID as a filename. Not to use ZIP or RAR. For example, your Moodle submission file name should look like this (two id numbers): `123456789_987654321.ipynb`.

### Question 1 - Data exploration and visialization - practical

Load Boston dataset from sklearn.
Explore the data. follow th instructions below and make sure to support your answers with proper outputs and plots.
When plotting, pay close attention to the range of the axis, and include axis labels and a title for the figure.

1. Describe the dataset. How many samples does it contain? How many features? What is the data type for each variable?
2. Produce a histogram and a boxplot of the nitric oxides concentration. describe the distribution.
3. Produce a correlation matrix of all the features. Are there any correlated features? Can you identify one feature with unusual behaviour?
4. Select the 2 pairs of features with the highest correlation (positive or negative) and plot 2 scatter plots with marginal histograms (JointPlot). 
5. Produce a cumulative histogram of the age variable and add two horizontal lines on the first and third quartile (on the cumulative count)
6. Identify and report 2 “interesting” trends in the data. No need to provide statistical confidence at this point. 

In [1]:
# <!-- <p style="color:red;"> -->
%pylab inline 
import pandas as pd

df=pd.read_...
# answer to 1 
df.info()
df.describe()
# answer to 2
df.nitric_oxide_concentration.hist(kind="hist")
df.nitric_oxide_concentration.plot(kind="box")

# answer to 3
corr_mat = df.corr()

# answer to 4
most_extreme_pair = corr_mat.stack().apply(np.abs).idxmax()
# some code that does joint plot.. 
# answer to 5 
age_cumsum = df.age.sort_values().cumsum()
age_cumsum.hist()
first_quartile = np.quantile(age_cumsum,0.25)
third_quartile = np.quantile(age_cumsum,0.75)
plt.plot([first_quartile]*2,[0,1],'r')
plt.plot([third_quartile]*2,[0,1],'r')
# answer to 6
# find trends by correlation, value counts. maybe we can copy some known trends from the web

SyntaxError: invalid syntax (1165197209.py, line 5)

### Question 2 - Independence and conditional independence

#### 2.A
Let $\ X, Y \ $ and $Z$  be discrete random variables with $\ n, m \ $ and $k=4$ possible outcomes respectivley.

How many parameters define the joint distribution of $\ X, Y \ $ and $Z$?

<p style="color:red;">
    Given no additional information, we need to know all possible outputs to to define the distribution and also all their probabilities (-1 because we have a constraint on the sum of probabilities). 
    which means we need (2n-1) + (2m-1) + (2k-1) parameters.
    The joint distribution is defined by the same amount of parameters. 

#### 2.B
For the same random variables from the previous section, how many parameters define the joint distribution of $\ X, Y \ $ and $Z$ if we now know that they are independent?

<p style="color:red;">
    Same answer. Knowing they are independent does not provide with additional constraints

#### 2.C
For the same random variables from the previous section, how many parameters define the joint distribution of $\ X, Y \ $ and $Z$ if we now know that $X$ and $Y$ are conditionaly independent given $Z$?

#### 2.D
Give an example for a joint distribution of $\ U, V \ $ and $W$ where $U$ and $V$ are NOT conditionally independent given $W$, but $U$ and $V$ are (unconditionally) independent. Also: $U$ and $V$ are standard normally distributed ($N(0, 1)$) and $W$ is a coin (2 possible values).

### Question 3 - Gaussian mixtures – parameter estimation and generation 

Consider the data provided in GMD.csv
Assume that the data comes from a Gaussian mixture distribution (GMD) with $k=3$. Furthermore, assume that $\mu_{1}=4, \mu_{2}=9, \sigma_{1}=\sigma_{2}=0.5, \sigma_3=1.5$ and $w_2=0.25$.

Read the data and answer the following questions.

#### 3.A
Provide an estimate for the other parameters of the distribution in two different ways.

<p style="color:red;">
    - 1) chi2 minimization (we have 3 unknown parameters) 
    - 2) depending on the data we might be able to see the peaks and get the weights fomr them (1 and 2 are well seperated (at least one of them is) so it should be possible)
    - 3) not sure what other ways there are... 
    - 4) EM algorithm

#### 3.B
Plot a graph of the pdf of the distribution you inferred. Select adequate limits for the axes for this plot and explain your decision.

<p style="color:red;">
    we just need to plot the function we retrieve.. 

#### 3.C
Now assume that the data comes from a Gaussian mixture distribution (GMD) with $k=4$.

The given data and parameters stay the same.

Can you estimate the unknown parameters in the two ways described in section A? Explain.

<p style="color:red;">
- 1) possible by adding 2 degrees of freedom and try to minimize chi2 farther?
- 2) only by guessing we in the data we see another peak. Probably a bad idea unless it will be very clear the peak is there. 

#### 3.D
Describe two ways for generating data for a GMM random variable with:
* centers at  $\mu_1=3, \mu_2=7, \mu_3=10$
* $\sigma_1=\sigma_2=\sigma_3=1$
* $w_1=w_2=w_3=0.33$

<p style="color:red;">
- 1) Monte Carlo (MC) - randomly choose x and pick it if some random value p bigger than the pdf?
- 2) Create 3 different distributions (using the same MC method) and then merge them together? In this case it is a simple combination because the weights are identical
- 3) Something other than MC

#### 3.E
Use one of the above approaches to generate 1000 points and plot a histogram of the result (decide on bins, axes etc)


#### 3.F
Use the other one to generate 1000 more points and draw two comparative histograms.

### Question 4 - Normally distributed salaries

The annual salaries of employees in a large Randomistan company are approximateley normally distributed with a mean of 75,000 RCU and a standard deviation of 37,500 RCU.

In [9]:
from scipy import stats # using this librarys implementation of the normal distribution

#### 4.A
What percent of people earn less than 50,000 RCU?

In [11]:
distribution = stats.norm(loc=75000,scale=37500)

<p style="color:red;">
fraction less than X is exactly the CDF, multiplying by 100 to get percents

In [13]:
distribution.cdf(50_000) * 100 

25.24925375469229

#### 4.B
What percent of people earn between 45,000 RCU and 65,000 RCU?

<p style="color:red;">
using the CDF to retrieve the integral in the range 45k-65k RCU:

In [24]:
(distribution.cdf(65_000) - distribution.cdf(45_000))*100  

18.30075118806284

#### 4.C
What percent of people earn more than 85,000 RCU?

<p style="color:red;">
using 1 - CDF to retrieve the inegration from 85k RCU to infinity:

In [19]:
(1-distribution.cdf(85_000)) * 100 

39.48629104640251

#### 4.D
The company has 1000 employees. How many employees in the company do you expect to earn more than 140,000 RCU?

In [23]:
(1-distribution.cdf(140_000)) * 1000

41.51821968877911

<p style="color:red;">
The calculation shows that there should be 41.5 people with earnings over 140k RCU, but because people count should be a natural number, we would expect the actual amount to be some natural number in the the vicinity of 41.5 (otherwise the salaries will not be approximately normally distributed)

### Question 5 - Coupon collector

Let $T_{N}$ denote the waiting time for full single coupon collection with N different equiprobable coupon types

#### 5.A
Write code to compute the exact value of $E(T_{N})$

In [32]:
import math

In [33]:
E_X_i=lambda i,N:N/(N-i)
E_T_N = lambda N:sum([E_X_i(i,N) for i in range(N)])

In [37]:
E_T_N(200) # exact value value

1175.6061896242886

In [38]:
200*math.log(200) # estimation from lectures (nH(n)~nlog(n), where H is the harmoinc sum up to n)

1059.6634733096073

#### 5.B
Write code to compute the exact value of $V(T_{N})$

In [42]:
V_X_i=lambda i,N:(1-(N-i)/N)/((N-i)/N)**2
V_T_N=lambda N:sum([V_X_i(i,N) for i in range(N)])

In [44]:
V_T_N(500) # exact value 

407337.60466372827

In [45]:
500**2*math.pi**2/6 - 500*math.log(500) # estimation from lectures (n^2*pi^2/6 - nH(n))

408126.21266284544

#### 5.C
Write code to exactly compute $P(T_{30}>60)$

In [107]:
def geometric(p, k):
    """
    probability for k coin tosses in a geometric distribution with coint p (failure rate)
    """
    if k == 0:
        return 0
    return (1 - p) ** (k - 1) * p


assert geometric(0.5, 0) == 0
assert geometric(0.5, 1) == 0.5
assert geometric(0.3, 2) == 0.3*0.7


def geometric_cdf(p, k):
    """
    probability for k coin tosses in a geometric distribution with coin p (failure rate)
    """
    if p==1: # edge case treatment
        return int(k==1)
    return p*(1-(1-p)**k)/(1-p)  # p(q + q^2 +...+q^k) = p(1-q^k)/q


assert geometric_cdf(0.5, 10) == sum([geometric(0.5, i) for i in range(11)])


def coupon(N, k, N_total=None):
    """
    probability to have k coupons before collecting all N variation
    N - coupon types left to collect
    k - coupons left to collect (aka try amount)
    N_orig - total coupon amount

    Note: implementation from the presentation in lecture 3 slide 21
    Note2: this method has complexity of O(k^N) (T(N) = kT(N-1)). 
    """
    if N_total is None:  # if not defined, we assume N is the total coupon amount
        N_total = N
    # stopping conditions
    if k < N or k == 0:  # not possible to achieve cases
        return 0
    if N == 1: # completing the last coupon
        return geometric_cdf(1/N_total, k) # if we want P(T_N=k) we can change geometric_cdf with geometric
    # convolution on N-1
    # (sum over probabilities for collecting N-1 coupons with some k' and then getting
    # the final coupon with k-k')
    prob_new = N / N_total  # probability to get a coupon we already saw
    conv_results = []
    for i in range(1, k):
        conv_part = geometric(prob_new, k - i) * coupon(N - 1, i, N_total)
        conv_results.append(conv_part)
    result = sum(conv_results)
    return result

assert coupon(2,2)==0.5
assert coupon(2,3)==0.75

def coupon_DP(N, k, N_total=None):
    """
    Same as coupon method but with dynamic programming implementation. 
    Note: This method has complexity O(Nk^2)
    :param N:
    :param k:
    :param N_total:
    :return:
    """
    if N_total is None:
        N_total = N
    if k==0:
        return 0
    M = [[0]*k for _ in range(N)] # N rows, k columns (starting from 1)
    for j in range(N-1,k):
        M[0][j] = geometric_cdf(1 / N_total, j)
    for n in range(1,N):
        for j in range(k):
            prob_new = N / N_total
            for i in range(j, k):
                M[n][j] = M[n][j] + geometric(prob_new, k - i) * M[n - 1][i]
    return M[N-1][k-1]


assert coupon_DP(2,5) == coupon(2,5)

In [111]:
[print(i,coupon_DP(30,i)) for i in range(30,160)]

30 0.02158163447818274
31 0.022011671949599635
32 0.02242737483863597
33 0.022829220964704425
34 0.023217672219903934
35 0.023593175099930124
36 0.023956161217288774
37 0.024307047797402136
38 0.024646238158178388
39 0.024974122173595428
40 0.025291076721831906
41 0.025597466118460494
42 0.025893642535201468
43 0.02617994640471774
44 0.0264567068119168
45 0.02672424187220923
46 0.02698285909715858
47 0.02723285574794295
48 0.027474519177034504
49 0.027708127158489675
50 0.027933948207229675
51 0.028152241887678342
52 0.028363259112112054
53 0.02856724242906464
54 0.028764426302118805
55 0.0289550373794045
56 0.029139294754114007
57 0.029317410216333196
58 0.029489588496478408
59 0.029656027500618784
60 0.029816918537954484
61 0.02997244654071232
62 0.030122790276711563
63 0.030268122554844165
64 0.03040861042370568
65 0.030544415363605148
66 0.030675693472174637
67 0.0308025956437918
68 0.03092526774302173
69 0.031043850772277323
70 0.03115848103389107
71 0.03126929028678436
72 0.03137

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

In [110]:
[print(i,coupon_DP(30,i,30) - coupon_DP(15,i,30)) for i in range(30,60)]

30 0.02158031723975023
31 0.022010328463762868
32 0.022426005980308417
33 0.02282782757963578
34 0.02321625512565223
35 0.023591735086801468
36 0.023954699049245725
37 0.024305564212941846
38 0.024644733871181428
39 0.024972597874146355
40 0.02528953307701246
41 0.02559590377311635
42 0.02589206211268345
43 0.02617834850759831
44 0.026455092022682675
45 0.026722610753930898
46 0.026981212194137516
47 0.02723119358633724
48 0.02747284226546364
49 0.027706435988619164
50 0.027932243254336166
51 0.02815052361119594
52 0.028361527956160388
53 0.02856549882295935
54 0.028762670660865013
55 0.028953270104173822
56 0.029137516232705676
57 0.029315620823619798
58 0.029487788594836778
59 0.02965421744034653


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

In [93]:

1 - coupon_DP(30,60)


0.9701830814620456

#### 5.D
Let $S_N$ denote the waiting time until $N/2$ different coupons are collected.

Write code to exactly compute $P(S_{30}>60)$

In [101]:
S_N_k=lambda N,k:coupon_DP(N//2,k,N)
1-S_N_k(30,60)

0.9999981801197182

In [102]:
S_N_k(30,60)

1.8198802818575735e-06